{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ec117f5d-ea8d-4378-8a2e-6d20c6309833",
   "metadata": {},
   "source": [
    "if (!require(\"BiocManager\", quietly = TRUE))\n",
    "    install.packages(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "e7a43a47-505b-4b6b-abd6-c7eb952816e5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.16 (BiocManager 1.30.19), R 4.2.2 (2022-10-31)\n",
      "\n",
      "Installing package(s) 'Biostrings'\n",
      "\n",
      "also installing the dependencies ‘bitops’, ‘zlibbioc’, ‘RCurl’, ‘GenomeInfoDbData’, ‘BiocGenerics’, ‘S4Vectors’, ‘IRanges’, ‘XVector’, ‘GenomeInfoDb’\n",
      "\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n"
     ]
    }
   ],
   "source": [
    "BiocManager::install(\"Biostrings\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "50c2a845-fbbd-4d80-b44b-f289efbbaa7c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "\n",
      "Attaching package: ‘Biostrings’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Biostrings’:\n",
      "\n",
      "    collapse, intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:XVector’:\n",
      "\n",
      "    slice\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(stringi)\n",
    "library(data.table)\n",
    "library(tools)\n",
    "library(data.table)\n",
    "library(stringdist)\n",
    "library(stringr)\n",
    "library(scales)\n",
    "library(ggplot2)\n",
    "library(Biostrings)\n",
    "library(stringr)\n",
    "library(dplyr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cd8f9f8f-ed79-4737-8f05-a24d242a5b74",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Set working directory \n",
    "setwd(\"/groups/nordborg/user/elizaveta.grigoreva/genome_evolution_pacbio\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "1c6e7f68-a05e-4826-8c28-f958c9b6356b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load data per chromosome (Anna's SNPs)\n",
    "chr1 <- readRDS('data/annas_snps/snps_mx1_final.rds')\n",
    "chr2 <- readRDS('data/annas_snps/snps_mx2_final.rds')\n",
    "chr3 <- readRDS('data/annas_snps/snps_mx3_final.rds')\n",
    "chr4 <- readRDS('data/annas_snps/snps_mx4_final.rds')\n",
    "chr5 <- readRDS('data/annas_snps/snps_mx5_final.rds')\n",
    "#Load indexies per chromosome (Anna's SNPs)\n",
    "chr1_idx <- as.data.frame(readRDS('data/annas_snps/snps_mx1_idx.rds'))\n",
    "chr2_idx <- as.data.frame(readRDS('data/annas_snps/snps_mx2_idx.rds'))\n",
    "chr3_idx <- as.data.frame(readRDS('data/annas_snps/snps_mx3_idx.rds'))\n",
    "chr4_idx <- as.data.frame(readRDS('data/annas_snps/snps_mx4_idx.rds'))\n",
    "chr5_idx <- as.data.frame(readRDS('data/annas_snps/snps_mx5_idx.rds'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3aca925f-94a6-4a1f-8861-d5296e6f837e",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf <- as.data.frame(cbind(chr1_idx = chr1_idx$`0`, chr1)) #Combine SNPs per individual and ids for Col0\n",
    "colnames(vcf)[1] <- \"POS\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "546e7911-f919-4c70-92bf-72085db11279",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Change chromosome notation per chromosome\n",
    "vcf <- as.data.frame(cbind(chr5_idx = chr5_idx$`0`, chr5)) #Combine SNPs per individual and ids for Col0\n",
    "colnames(vcf)[1] <- \"POS\"\n",
    "\n",
    "colnames(vcf)[2] <- \"REF\"\n",
    "#Add ALT allele frequencies (non-reference)\n",
    "uniq_values <- apply(vcf[,3:29],1,unique) #From 3(first sample) to the last ncol(vcf)\n",
    "###Transform  list to the matrix to get a list of alternative alleles \n",
    "ALT <- stri_list2matrix(uniq_values, byrow=TRUE) #Chr1\n",
    "#Make a dataframe of all alternative alleles\n",
    "colnames(ALT) <- c(\"ALT1\", \"ALT2\", \"ALT3\",\"ALT4\", \"ALT5\")  #Number of alleles based on the number of columns\n",
    "#Insert reference values to each column\n",
    "for(i in 3:ncol(vcf)) {       # for-loop over columns\n",
    "  vcf[ , i] <- paste(vcf$REF,vcf[ , i],sep=\"\")}\n",
    "#We have 1st allele always REF. If first allele present but alternative is lost, we will keep this site as NA\n",
    "vcf <- data.frame(lapply(vcf, function(x) {\n",
    "  gsub(\"cNA\", \"NA\",\n",
    "  gsub(\"aNA\", \"NA\",\n",
    "  gsub(\"gNA\", \"NA\",\n",
    "  gsub(\"tNA\", \"NA\" , x))))\n",
    "  }))\n",
    "#Add alternative allele \n",
    "ALT <- as.data.frame(ALT)\n",
    "vcf$ALT <- paste0(ALT$ALT1,\",\",ALT$ALT2,\",\",ALT$ALT3,\",\",ALT$ALT4,\",\", ALT$ALT5 )\n",
    "##Remove NA values that represent absence of ALT allele.We are not interested in casese when reference allele is present and alt are not \n",
    "#Remove NAs\n",
    "vcf$ALT <- gsub(\"NA,\",\"\",vcf$ALT)\n",
    "vcf$ALT <- gsub(\",NA,\",\"\", vcf$ALT)\n",
    "vcf$ALT <- gsub(\",NA\",\"\", vcf$ALT)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "03da8586-4705-4a7c-b4f5-16894ffa3ed2",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Make the file Hapmap like\n",
    "###Add hapmap-like columns\n",
    "vcf$rs<- seq(1:nrow(vcf)) #Insert numeric identifier of SNP\n",
    "vcf$alleles <- paste(vcf$REF,vcf$ALT,sep=\",\") #Insert reference and alternative allele\n",
    "vcf$alleles <- sapply(strsplit(vcf$alleles,','), function(x) toString(unique(x))) #Make alleles unique (calculate eneterez of REF+ALT unque combinations, because we are interested in variations)! Keep monomorphic sites for PI calculation\n",
    "vcf$counts <- nchar(vcf$alleles)#Calculate number of characters\n",
    "vcf <- vcf[vcf$counts==4,] #Filter only bi-allelic variants 4 characters: letter, del, tab, letter)\n",
    "vcf$alleles <-  str_replace(vcf$alleles , ', ', '/') #Change delimeter in alleles\n",
    "\n",
    "###Make HapMap-like file\n",
    "vcf$chrom <- '5' #Change it for each chromosome \n",
    "vcf$pos <- vcf$POS\n",
    "vcf$strand <- \".\"\n",
    "vcf$center <- \".\"\n",
    "vcf$protLSID <- \".\"\n",
    "vcf$assayLSID <- \".\"\n",
    "vcf$panelLSID <- \".\"\n",
    "vcf$QCcode <- \".\"\n",
    "vcf$assembly <- \".\"\n",
    "\n",
    "###Drop unnecessary columns\n",
    "vcf$CHROM <- NULL\n",
    "vcf$POS <- NULL\n",
    "vcf$REF <- NULL\n",
    "vcf$ALT <- NULL\n",
    "vcf$counts <- NULL\n",
    "##Transpose columns \n",
    "vcf <- vcf %>% relocate(rs)\n",
    "vcf <- vcf %>% relocate(alleles, .after =  rs)\n",
    "vcf <- vcf %>% relocate(chrom, .after =  alleles)\n",
    "vcf <- vcf %>% relocate(pos, .after =  chrom)\n",
    "vcf <- vcf %>% relocate(strand, .after =  pos)\n",
    "vcf <- vcf %>% relocate(assembly, .after =  strand)\n",
    "vcf <- vcf %>% relocate(center, .after =  assembly)\n",
    "vcf <- vcf %>% relocate(protLSID, .after =  center)\n",
    "vcf <- vcf %>% relocate(assayLSID, .after =  protLSID)\n",
    "vcf <- vcf %>% relocate(panelLSID, .after =  assayLSID)\n",
    "vcf <- vcf %>% relocate(QCcode, .after =  panelLSID)\n",
    "\n",
    "\n",
    "###Change small letters to big and NA to gap notation\n",
    "vcf <- data.frame(lapply(vcf, function(x) {\n",
    "  gsub(\"NA\", \"NN\",\n",
    "  gsub(\"t\", \"T\",\n",
    "  gsub(\"a\", \"A\",\n",
    "  gsub(\"c\", \"C\",\n",
    "  gsub(\"g\", \"G\",x)))))\n",
    "  }))\n",
    "                      \n",
    "\n",
    "#Remove unrecognized letters \n",
    "vcf_corr <- vcf[!grepl(\"k|m|r|w|y|n|s|b|h|v|d\",vcf$alleles ),]\n",
    "vcf_corr$pos <- as.integer(vcf_corr$pos) #Make position integer value \n",
    "###Write to the file\n",
    "write.table(vcf_corr,\"/groups/nordborg/user/elizaveta.grigoreva/genome_evolution_pacbio/analyses/004_polymorphism_annas_snps/hmp_parsed/chr5_27acc.hmp\",quote=FALSE, row.names = F, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "1a7e74e2-0c43-455a-8580-f78227a71c15",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 5 × 38</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>rs</th><th scope=col>alleles</th><th scope=col>chrom</th><th scope=col>pos</th><th scope=col>strand</th><th scope=col>assembly</th><th scope=col>center</th><th scope=col>protLSID</th><th scope=col>assayLSID</th><th scope=col>panelLSID</th><th scope=col>QCcode</th><th scope=col>X22002</th><th scope=col>X9537</th><th scope=col>X9905</th><th scope=col>X10002</th><th scope=col>X6966</th><th scope=col>X10024</th><th scope=col>X6069</th><th scope=col>X6124</th><th scope=col>X6244</th><th scope=col>X220011</th><th scope=col>X22003</th><th scope=col>X6024</th><th scope=col>X9075</th><th scope=col>X9638</th><th scope=col>X9981</th><th scope=col>X22004</th><th scope=col>X22005</th><th scope=col>X22006</th><th scope=col>X9543</th><th scope=col>X10015</th><th scope=col>X1741</th><th scope=col>X8236</th><th scope=col>X9728</th><th scope=col>X9888</th><th scope=col>X9764</th><th scope=col>X22007</th><th scope=col>X6909</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>1</td><td>C/T</td><td>1</td><td> 1</td><td>.</td><td>.</td><td>.</td><td>.</td><td>.</td><td>.</td><td>.</td><td>CT</td><td>NN</td><td>NN</td><td>CC</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>CC</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>2</td><td>A/T</td><td>1</td><td> 7</td><td>.</td><td>.</td><td>.</td><td>.</td><td>.</td><td>.</td><td>.</td><td>AT</td><td>NN</td><td>NN</td><td>AA</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>AA</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>3</td><td>C/A</td><td>1</td><td> 9</td><td>.</td><td>.</td><td>.</td><td>.</td><td>.</td><td>.</td><td>.</td><td>CA</td><td>NN</td><td>NN</td><td>CC</td><td>CC</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>CC</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>4</td><td>A/T</td><td>1</td><td>12</td><td>.</td><td>.</td><td>.</td><td>.</td><td>.</td><td>.</td><td>.</td><td>AT</td><td>NN</td><td>NN</td><td>AA</td><td>AA</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>AA</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>5</td><td>C/T</td><td>1</td><td>15</td><td>.</td><td>.</td><td>.</td><td>.</td><td>.</td><td>.</td><td>.</td><td>CT</td><td>NN</td><td>NN</td><td>CC</td><td>CC</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>CC</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td><td>NN</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 5 × 38\n",
       "\\begin{tabular}{r|llllllllllllllllllllllllllllllllllllll}\n",
       "  & rs & alleles & chrom & pos & strand & assembly & center & protLSID & assayLSID & panelLSID & QCcode & X22002 & X9537 & X9905 & X10002 & X6966 & X10024 & X6069 & X6124 & X6244 & X220011 & X22003 & X6024 & X9075 & X9638 & X9981 & X22004 & X22005 & X22006 & X9543 & X10015 & X1741 & X8236 & X9728 & X9888 & X9764 & X22007 & X6909\\\\\n",
       "  & <chr> & <chr> & <chr> & <int> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & 1 & C/T & 1 &  1 & . & . & . & . & . & . & . & CT & NN & NN & CC & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & CC & NN & NN & NN & NN & NN\\\\\n",
       "\t2 & 2 & A/T & 1 &  7 & . & . & . & . & . & . & . & AT & NN & NN & AA & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & AA & NN & NN & NN & NN & NN\\\\\n",
       "\t3 & 3 & C/A & 1 &  9 & . & . & . & . & . & . & . & CA & NN & NN & CC & CC & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & CC & NN & NN & NN & NN & NN\\\\\n",
       "\t4 & 4 & A/T & 1 & 12 & . & . & . & . & . & . & . & AT & NN & NN & AA & AA & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & AA & NN & NN & NN & NN & NN\\\\\n",
       "\t5 & 5 & C/T & 1 & 15 & . & . & . & . & . & . & . & CT & NN & NN & CC & CC & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & NN & CC & NN & NN & NN & NN & NN\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 5 × 38\n",
       "\n",
       "| <!--/--> | rs &lt;chr&gt; | alleles &lt;chr&gt; | chrom &lt;chr&gt; | pos &lt;int&gt; | strand &lt;chr&gt; | assembly &lt;chr&gt; | center &lt;chr&gt; | protLSID &lt;chr&gt; | assayLSID &lt;chr&gt; | panelLSID &lt;chr&gt; | QCcode &lt;chr&gt; | X22002 &lt;chr&gt; | X9537 &lt;chr&gt; | X9905 &lt;chr&gt; | X10002 &lt;chr&gt; | X6966 &lt;chr&gt; | X10024 &lt;chr&gt; | X6069 &lt;chr&gt; | X6124 &lt;chr&gt; | X6244 &lt;chr&gt; | X220011 &lt;chr&gt; | X22003 &lt;chr&gt; | X6024 &lt;chr&gt; | X9075 &lt;chr&gt; | X9638 &lt;chr&gt; | X9981 &lt;chr&gt; | X22004 &lt;chr&gt; | X22005 &lt;chr&gt; | X22006 &lt;chr&gt; | X9543 &lt;chr&gt; | X10015 &lt;chr&gt; | X1741 &lt;chr&gt; | X8236 &lt;chr&gt; | X9728 &lt;chr&gt; | X9888 &lt;chr&gt; | X9764 &lt;chr&gt; | X22007 &lt;chr&gt; | X6909 &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 1 | C/T | 1 |  1 | . | . | . | . | . | . | . | CT | NN | NN | CC | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | CC | NN | NN | NN | NN | NN |\n",
       "| 2 | 2 | A/T | 1 |  7 | . | . | . | . | . | . | . | AT | NN | NN | AA | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | AA | NN | NN | NN | NN | NN |\n",
       "| 3 | 3 | C/A | 1 |  9 | . | . | . | . | . | . | . | CA | NN | NN | CC | CC | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | CC | NN | NN | NN | NN | NN |\n",
       "| 4 | 4 | A/T | 1 | 12 | . | . | . | . | . | . | . | AT | NN | NN | AA | AA | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | AA | NN | NN | NN | NN | NN |\n",
       "| 5 | 5 | C/T | 1 | 15 | . | . | . | . | . | . | . | CT | NN | NN | CC | CC | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | NN | CC | NN | NN | NN | NN | NN |\n",
       "\n"
      ],
      "text/plain": [
       "  rs alleles chrom pos strand assembly center protLSID assayLSID panelLSID\n",
       "1 1  C/T     1      1  .      .        .      .        .         .        \n",
       "2 2  A/T     1      7  .      .        .      .        .         .        \n",
       "3 3  C/A     1      9  .      .        .      .        .         .        \n",
       "4 4  A/T     1     12  .      .        .      .        .         .        \n",
       "5 5  C/T     1     15  .      .        .      .        .         .        \n",
       "  QCcode X22002 X9537 X9905 X10002 X6966 X10024 X6069 X6124 X6244 X220011\n",
       "1 .      CT     NN    NN    CC     NN    NN     NN    NN    NN    NN     \n",
       "2 .      AT     NN    NN    AA     NN    NN     NN    NN    NN    NN     \n",
       "3 .      CA     NN    NN    CC     CC    NN     NN    NN    NN    NN     \n",
       "4 .      AT     NN    NN    AA     AA    NN     NN    NN    NN    NN     \n",
       "5 .      CT     NN    NN    CC     CC    NN     NN    NN    NN    NN     \n",
       "  X22003 X6024 X9075 X9638 X9981 X22004 X22005 X22006 X9543 X10015 X1741 X8236\n",
       "1 NN     NN    NN    NN    NN    NN     NN     NN     NN    NN     NN    CC   \n",
       "2 NN     NN    NN    NN    NN    NN     NN     NN     NN    NN     NN    AA   \n",
       "3 NN     NN    NN    NN    NN    NN     NN     NN     NN    NN     NN    CC   \n",
       "4 NN     NN    NN    NN    NN    NN     NN     NN     NN    NN     NN    AA   \n",
       "5 NN     NN    NN    NN    NN    NN     NN     NN     NN    NN     NN    CC   \n",
       "  X9728 X9888 X9764 X22007 X6909\n",
       "1 NN    NN    NN    NN     NN   \n",
       "2 NN    NN    NN    NN     NN   \n",
       "3 NN    NN    NN    NN     NN   \n",
       "4 NN    NN    NN    NN     NN   \n",
       "5 NN    NN    NN    NN     NN   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Convert to vcf file \n",
    "run_pipeline.pl -Xmx5g -fork1 -h chr2_27acc.hmp -export -exportType VCF -runfork1\n",
    "#Check reference mismatches \n",
    "bcftools norm --check-ref s -f GCF_000001735.4_TAIR10.1_genomic_5Chr_renamd_for_syri.fna chr2_27acc.vcf  -o  chr2_27acc_corrected.vcf\n",
    "#Calculate PI using vcftools\n",
    "vcftools --vcf chr1_27acc_corrected.vcf --window-pi  100000 --out chr1_27acc_corrected_pi_100KB"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R [conda env:.conda-R]",
   "language": "R",
   "name": "conda-env-.conda-R-r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
